File: backend/agents/SupplierAgent.js
// backend/agents/SupplierAgent.js
const { subscribe, publish } = require("../eventBus");
const EVENTS = require("../constants/events");

class SupplierAgent {
  constructor(id, worldState, log) {
    this.id = id;
    this.worldState = worldState;
    this.log = log;
    this.ML_SERVICE_URL = "http://localhost:8000"; // Python ML service

    subscribe("MEDICINE_SHORTAGE_RISK", this.onMedicineShortage.bind(this));
    subscribe("HOSPITAL_OVERLOAD_RISK", this.onHospitalOverload.bind(this));
    subscribe("EQUIPMENT_SHORTAGE", this.onEquipmentShortage.bind(this));
    subscribe("HOSPITAL_RESOURCE_REQUEST", this.onResourceRequest.bind(this));
  }

  start() {
    // Log initialization
    const s = this.worldState.suppliers[this.id];
    const itemCount = Object.keys(s.inventory || {}).length;
    const vehicleCount = Object.keys(s.fleet || {}).length;
    this.log(
      `✅ Supplier Agent ${this.id} (${s.name}) initialized - ${itemCount} items in warehouse | ${vehicleCount} vehicles`,
      { agent: "Supplier", type: "INIT", entityId: this.id }
    );

    // Periodic inventory check every 15 seconds (faster for demo)
    setInterval(() => this.tick(), 15000);
  }

  async tick() {
    const s = this.worldState.suppliers[this.id];
    if (!s) return;

    // Count low inventory items
    const lowStockItems = Object.entries(s.inventory).filter(([_, item]) => {
      const stock = item.stock || item;
      return typeof stock === "number" && stock < 100;
    }).length;

    // Count active orders
    const activeOrders = s.activeOrders?.length || 0;

    // Count available vehicles
    const availableVehicles = s.fleet
      ? Object.values(s.fleet).filter((v) => v.status === "available").length
      : 0;
    const totalVehicles = s.fleet ? Object.keys(s.fleet).length : 0;

    // ALWAYS log current status
    const status = activeOrders > 2 ? "🟡 BUSY" : "🟢 READY";
    this.log(
      `${s.name}: ${status} | ${activeOrders} active orders | ${lowStockItems} low stock alerts | ${availableVehicles}/${totalVehicles} vehicles ready`,
      {
        agent: "Supplier",
        type: "STATUS",
        entityId: this.id,
        activeOrders,
        lowStockItems,
      }
    );

    // Check for low inventory levels
    this.checkInventoryLevels(s);

    // 🤖 ML-POWERED: Process orders with intelligent prioritization
    await this.processActiveOrdersWithML(s);
  }

  checkInventoryLevels(supplier) {
    Object.keys(supplier.inventory).forEach((item) => {
      const itemData = supplier.inventory[item];
      const stock = itemData.stock || itemData; // Handle both object and number formats

      if (typeof stock === "number" && stock < 100) {
        this.log(
          `[Supplier ${this.id}] Low inventory warning: ${item} = ${stock} units`,
          { agent: "Supplier", type: "LOW_INVENTORY", item, stock }
        );
      }
    });
  }

  async processActiveOrdersWithML(supplier) {
    /**
     * 🤖 ML-POWERED ORDER PRIORITIZATION
     * Uses Python FastAPI ML service for intelligent order fulfillment
     * Formula: Priority_Score = (Strain×0.4) + (Criticality×0.3) + (Urgency×0.3)
     */

    if (!supplier.activeOrders || supplier.activeOrders.length === 0) return;

    try {
      // Prepare orders for ML service
      const orders = supplier.activeOrders.map((order) => ({
        order_id: order.orderId || `ORD-${Date.now()}`,
        requester_id: order.pharmacyId || order.hospitalId || "unknown",
        requester_type: order.hospitalId ? "hospital" : "pharmacy",
        medicine: order.medicine || order.item,
        quantity: order.quantity || 0,
        urgency: order.urgency || "medium",
        criticality: order.criticality || "medium",
        zone: order.zone || "unknown",
        requester_strain: order.requesterStrain || 0.5, // Default 50% if not provided
      }));

      // Call Python ML service
      const response = await fetch(`${this.ML_SERVICE_URL}/prioritize/orders`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          supplier_id: this.id,
          orders: orders,
          available_inventory: supplier.inventory,
        }),
      });

      if (!response.ok) {
        throw new Error(`ML Service error: ${response.status}`);
      }

      const result = await response.json();

      // Log ML prediction
      this.log(
        `🤖 ML ORDER PRIORITIZATION: ${supplier.name} | ${
          result.total_orders
        } orders | ${
          result.fulfilled_count
        } fulfilled (${result.fulfillment_rate.toFixed(1)}%)`,
        {
          agent: "Supplier",
          type: `ML_ORDER_PRIORITY`,
          entityId: this.id,
          totalOrders: result.total_orders,
          fulfilled: result.fulfilled_count,
          fulfillmentRate: result.fulfillment_rate,
          mlPowered: true,
        }
      );

      // Process prioritized orders
      result.prioritized_orders.forEach((prioritizedOrder, index) => {
        if (prioritizedOrder.fulfillment_status === "fulfilled") {
          this.log(
            `✅ ${supplier.name}: Order #${
              index + 1
            } (Priority: ${prioritizedOrder.priority_score.toFixed(1)}) - ${
              prioritizedOrder.medicine
            } to ${
              prioritizedOrder.requester_id
            } | Status: ${prioritizedOrder.fulfillment_status.toUpperCase()}`,
            {
              agent: "Supplier",
              type: `ML_ORDER_FULFILLED`,
              entityId: this.id,
              orderId: prioritizedOrder.order_id,
              priority: prioritizedOrder.priority_score,
              medicine: prioritizedOrder.medicine,
            }
          );
        } else {
          this.log(
            `⏳ ${supplier.name}: Order #${
              index + 1
            } (Priority: ${prioritizedOrder.priority_score.toFixed(1)}) - ${
              prioritizedOrder.medicine
            } | Status: ${prioritizedOrder.fulfillment_status.toUpperCase()} - ${
              prioritizedOrder.reason
            }`,
            {
              agent: "Supplier",
              type: `ML_ORDER_PENDING`,
              entityId: this.id,
              orderId: prioritizedOrder.order_id,
              priority: prioritizedOrder.priority_score,
              reason: prioritizedOrder.reason,
            }
          );
        }
      });
    } catch (error) {
      // Fallback: Use rule-based prioritization if ML service unavailable
      this.log(
        `⚠️ ML service unavailable for ${supplier.name}, using fallback prioritization: ${error.message}`,
        {
          agent: "Supplier",
          type: "ML_FALLBACK",
          entityId: this.id,
          error: error.message,
        }
      );
      this.processActiveOrders(supplier);
    }
  }

  processActiveOrders(supplier) {
    if (supplier.activeOrders && supplier.activeOrders.length > 0) {
      // Sort orders by priority (urgency and criticality)
      const prioritizedOrders = this.prioritizeOrders(supplier.activeOrders);

      // Process top priority orders
      prioritizedOrders.slice(0, 3).forEach((order) => {
        this.fulfillOrder(supplier, order);
      });
    }
  }

  prioritizeOrders(orders) {
    return orders.sort((a, b) => {
      // Calculate priority score
      const scoreA = this.calculateOrderPriority(a);
      const scoreB = this.calculateOrderPriority(b);
      return scoreB - scoreA; // Higher score first
    });
  }

  calculateOrderPriority(order) {
    let score = 0;

    // Urgency scoring
    if (order.urgency === "high") score += 50;
    else if (order.urgency === "medium") score += 30;
    else score += 10;

    // Criticality scoring
    if (order.criticality === "high") score += 30;
    else if (order.criticality === "medium") score += 15;

    // Zone risk (if available)
    if (order.zoneRisk === "high") score += 20;

    // From hospital vs pharmacy (hospitals get priority)
    if (order.hospitalId) score += 15;

    return score;
  }

  onResourceRequest(event) {
    const s = this.worldState.suppliers[this.id];
    if (!s) return;

    const {
      hospitalId,
      hospitalName,
      zone,
      recommendations,
      hsi,
      strainLevel,
      mlPowered,
    } = event;

    this.log(
      `📥 ${
        s.name
      }: ${strainLevel} PRIORITY resource request from ${hospitalName} (HSI: ${
        hsi?.toFixed(1) || "N/A"
      }) - ${recommendations?.join(", ") || "General supplies"}`,
      {
        agent: "Supplier",
        type: "RESOURCE_REQUEST",
        entityId: this.id,
        hospitalId,
        zone,
        strainLevel,
      }
    );

    // Add to active orders with high priority
    if (recommendations && recommendations.length > 0) {
      recommendations.forEach((recommendation) => {
        // Extract resource type from recommendation text
        let resourceType = "medical_supplies";
        if (recommendation.toLowerCase().includes("bed")) resourceType = "beds";
        else if (recommendation.toLowerCase().includes("staff"))
          resourceType = "staff";
        else if (recommendation.toLowerCase().includes("equipment"))
          resourceType = "equipment";

        const order = {
          orderId: `HSP-${Date.now()}-${Math.random()
            .toString(36)
            .substr(2, 9)}`,
          hospitalId,
          item: resourceType,
          quantity: 10, // Default quantity
          zone,
          urgency:
            strainLevel === "CRITICAL"
              ? "high"
              : strainLevel === "ELEVATED"
              ? "medium"
              : "low",
          criticality: "high",
          requesterStrain: hsi ? hsi / 100 : 0.6,
          status: "pending",
          timestamp: new Date().toISOString(),
          mlPowered,
        };

        s.activeOrders = s.activeOrders || [];
        s.activeOrders.push(order);
      });
    }
  }

  onMedicineShortage(event) {
    const s = this.worldState.suppliers[this.id];
    if (!s) return;

    const {
      medicine,
      zone,
      pharmacyId,
      pharmacyName,
      stock,
      orderQuantity,
      urgency,
      criticality,
      supplier,
    } = event;

    // Check if we're the designated supplier for this pharmacy
    if (supplier && supplier !== this.id) return;

    this.log(
      `📥 ${s.name}: ${
        urgency?.toUpperCase() || "NORMAL"
      } PRIORITY order received from ${
        pharmacyName || "Pharmacy"
      } - ${medicine} (${orderQuantity} units needed)`,
      {
        agent: "Supplier",
        type: "ORDER_RECEIVED",
        entityId: this.id,
        pharmacyId,
        medicine,
        zone,
        urgency,
      }
    );

    const itemData = s.inventory[medicine];
    if (!itemData) {
      this.log(
        `❌ ${s.name}: Cannot fulfill order - ${medicine} not in our inventory catalog`,
        {
          agent: "Supplier",
          type: "NOT_AVAILABLE",
          entityId: this.id,
          pharmacyId,
          medicine,
          zone,
        }
      );
      return;
    }

    const available = itemData.stock || itemData;
    if (available <= 0) {
      this.log(
        `❌ ${s.name}: Cannot supply ${medicine} to ${
          pharmacyName || "Pharmacy"
        } - OUT OF STOCK`,
        {
          agent: "Supplier",
          type: "NO_SUPPLY",
          entityId: this.id,
          pharmacyId,
          medicine,
          zone,
        }
      );

      publish("SUPPLY_UNAVAILABLE", {
        supplierId: this.id,
        supplierName: s.name,
        pharmacyId,
        medicine,
        zone,
      });
      return;
    }

    // Determine quantity to send
    const requestedQty = orderQuantity || Math.max(stock * 2, 50);
    const quantity = Math.min(available, requestedQty);

    // Update inventory
    if (typeof itemData === "object") {
      itemData.stock -= quantity;
    } else {
      s.inventory[medicine] -= quantity;
    }

    // Calculate ETA based on delivery fleet
    const eta = this.calculateDeliveryETA(s, zone);

    const remainingStock = itemData.stock || s.inventory[medicine];
    this.log(
      `✅ ${s.name}: Order CONFIRMED - Dispatching ${quantity} units of ${medicine} to ${zone} | ETA: ${eta}h | Warehouse remaining: ${remainingStock} units`,
      {
        agent: "Supplier",
        type: "SUPPLY_CONFIRMED",
        entityId: this.id,
        pharmacyId,
        medicine,
        quantity,
        zone,
        urgency,
        eta,
      }
    );

    // Log coordination message
    this.log(
      `📡 ${s.name}: Confirming delivery to ${
        pharmacyName || "pharmacy"
      } - ${quantity} units of ${medicine} dispatched to ${zone}`,
      {
        agent: "Supplier",
        type: "COORDINATION",
        entityId: this.id,
        recipient: pharmacyName || pharmacyId,
        medicine,
        quantity,
        zone,
      }
    );

    // Add to active orders
    const order = {
      orderId: `ORD-${Date.now()}`,
      pharmacyId,
      medicine,
      quantity,
      zone,
      urgency,
      criticality,
      status: "dispatched",
      timestamp: new Date().toISOString(),
      eta: `${eta} hours`,
    };
    s.activeOrders.push(order);

    // Publish confirmation
    publish("SUPPLY_CONFIRMED", {
      supplierId: this.id,
      pharmacyId,
      medicine,
      quantity,
      zone,
      eta: `${eta} hours`,
      orderId: order.orderId,
    });

    // Allocate delivery vehicle
    if (s.fleet && s.fleet.available > 0) {
      s.fleet.available--;
      s.fleet.inTransit++;

      // Simulate delivery completion (in real system, this would be actual tracking)
      setTimeout(() => {
        this.completeDelivery(s, order);
      }, eta * 1000); // Simulate in seconds for demo (normally would be hours)
    }
  }

  calculateDeliveryETA(supplier, zone) {
    if (!supplier.fleet) return 2; // Default 2 hours

    const baseTime = supplier.fleet.avgDeliveryTime || 2;
    // Add delay if fleet is busy
    const utilizationFactor =
      supplier.fleet.inTransit / supplier.fleet.vehicles;
    return baseTime + utilizationFactor * 0.5;
  }

  completeDelivery(supplier, order) {
    // Return vehicle to available pool
    if (supplier.fleet) {
      supplier.fleet.available++;
      supplier.fleet.inTransit--;
    }

    // Update order status
    const orderIndex = supplier.activeOrders.findIndex(
      (o) => o.orderId === order.orderId
    );
    if (orderIndex >= 0) {
      supplier.activeOrders[orderIndex].status = "delivered";

      // Remove after 1 minute (keep for reference briefly)
      setTimeout(() => {
        supplier.activeOrders.splice(orderIndex, 1);
      }, 60000);
    }

    this.log(
      `[Supplier ${this.id}] Delivery completed: ${order.medicine} to Pharmacy ${order.pharmacyId}`,
      { agent: "Supplier", type: "DELIVERY_COMPLETE", orderId: order.orderId }
    );

    publish("DELIVERY_COMPLETE", {
      supplierId: this.id,
      orderId: order.orderId,
      pharmacyId: order.pharmacyId,
      medicine: order.medicine,
    });
  }

  fulfillOrder(supplier, order) {
    // Process orders from activeOrders array
    if (order.status === "dispatched" || order.status === "delivered") return;

    // Similar logic to onMedicineShortage
    const itemData = supplier.inventory[order.medicine];
    if (!itemData) return;

    const available = itemData.stock || itemData;
    if (available <= 0) return;

    const quantity = Math.min(available, order.quantity);

    if (typeof itemData === "object") {
      itemData.stock -= quantity;
    } else {
      supplier.inventory[order.medicine] -= quantity;
    }

    order.status = "dispatched";
    order.actualQuantity = quantity;
  }

  onHospitalOverload(event) {
    const s = this.worldState.suppliers[this.id];
    if (!s) return;

    const { hospitalId, occupancy, zone, name } = event;

    // Check if we have critical equipment available
    const ventilators = s.inventory.ventilators;
    const hasVentilators =
      ventilators && (ventilators.stock || ventilators) > 0;

    if (hasVentilators) {
      this.log(
        `[Supplier ${this.id}] Hospital ${
          name || hospitalId
        } overload detected (${Math.round(
          occupancy * 100
        )}% occupancy). Ventilators available for priority supply.`,
        {
          agent: "Supplier",
          type: "OVERLOAD_RESPONSE",
          hospitalId,
          zone,
          hasVentilators,
        }
      );
    } else {
      this.log(
        `[Supplier ${this.id}] Hospital ${
          name || hospitalId
        } overload noted but no ventilators available.`,
        { agent: "Supplier", type: "OVERLOAD_NOTICE", hospitalId, zone }
      );
    }
  }

  onEquipmentShortage(event) {
    const s = this.worldState.suppliers[this.id];
    if (!s) return;

    const { hospitalId, zone, equipment, available, total } = event;

    const equipmentData = s.inventory[equipment];
    if (!equipmentData) return;

    const stock = equipmentData.stock || equipmentData;
    if (stock > 0) {
      // Can supply equipment
      const quantityToSend = Math.min(
        stock,
        Math.ceil((total - available) * 0.5)
      );

      this.log(
        `[Supplier ${this.id}] Critical ${equipment} shortage at Hospital ${hospitalId}. Preparing ${quantityToSend} units for priority delivery.`,
        {
          agent: "Supplier",
          type: "EQUIPMENT_SUPPLY",
          hospitalId,
          equipment,
          quantity: quantityToSend,
        }
      );

      // In a real system, this would trigger equipment delivery
      if (typeof equipmentData === "object") {
        equipmentData.stock -= quantityToSend;
      } else {
        s.inventory[equipment] -= quantityToSend;
      }
    }
  }
}

module.exports = SupplierAgent;

File: backend/ml_service/agents/supplier_agent.py
"""
Supplier Agent - Supply Chain Optimization

Implementation Mandate: Hybrid Logic
- Formula: Calculate a Priority Score for all incoming orders:
  Priority_Score = (Requester_Strain * 0.4) + (Medicine_Criticality * 0.3) + 
                   (Requested_Urgency * 0.3)
  
- Rule: Fulfill strictly by Priority Score (highest first) and log the action
"""

from typing import Dict, List, Optional
from datetime import datetime


class SupplierAgent:
    """Supplier Agent for supply chain management and order prioritization"""
    
    def __init__(self):
        self.URGENCY_WEIGHTS = {
            'URGENT': 100,
            'HIGH': 75,
            'MEDIUM': 50,
            'NORMAL': 25,
            'LOW': 10
        }
        
        self.MEDICINE_CRITICALITY = {
            # Life-saving medicines
            'oxygen': 100,
            'iv_fluids': 95,
            'antibiotics': 90,
            'covid_medicine': 90,
            'dengue_medicine': 85,
            'malaria_medicine': 85,
            'typhoid_medicine': 85,
            
            # Important medicines
            'antimalarial': 75,
            'antivirals': 75,
            'flu_medicine': 70,
            'paracetamol': 60,
            
            # General supplies
            'syringes': 50,
            'bandages': 40,
            'surgical_masks': 40,
            'gloves': 35,
            'default': 50  # For unspecified medicines
        }
    
    def prioritize_orders(
        self,
        orders: List[Dict],  # List of incoming orders
        inventory: Dict[str, int],  # Current warehouse inventory
        delivery_capacity: int = 4  # Number of available delivery vehicles
    ) -> Dict:
        """
        Prioritize and fulfill orders based on Priority Score
        
        Priority Score Formula:
        PS = (Requester_Strain * 0.4) + (Medicine_Criticality * 0.3) + (Urgency * 0.3)
        
        Orders are fulfilled strictly by highest priority first
        """
        
        prioritized_orders = []
        fulfilled_orders = []
        pending_orders = []
        
        # Calculate priority score for each order
        for order in orders:
            priority_score = self._calculate_priority_score(
                requester_strain=order.get('requester_strain', 50),
                medicine=order.get('medicine', 'default'),
                urgency=order.get('urgency', 'NORMAL'),
                quantity=order.get('quantity', 0)
            )
            
            order_with_priority = {
                **order,
                'priority_score': priority_score,
                'timestamp': order.get('timestamp', datetime.now().isoformat())
            }
            
            prioritized_orders.append(order_with_priority)
        
        # Sort by priority score (highest first)
        prioritized_orders.sort(key=lambda x: x['priority_score'], reverse=True)
        
        # Fulfill orders based on priority and inventory availability
        available_vehicles = delivery_capacity
        
        for order in prioritized_orders:
            medicine = order.get('medicine', 'default')
            quantity = order.get('quantity', 0)
            requester = order.get('requester_id', 'UNKNOWN')
            
            # Check inventory availability
            available_stock = inventory.get(medicine, 0)
            
            if available_stock >= quantity and available_vehicles > 0:
                # Can fulfill order
                fulfillment = {
                    **order,
                    'status': 'FULFILLED',
                    'allocated_quantity': quantity,
                    'fulfillment_time': datetime.now().isoformat(),
                    'estimated_delivery': '4-8 hours' if order.get('urgency') == 'URGENT' else '24 hours'
                }
                fulfilled_orders.append(fulfillment)
                
                # Update inventory and vehicle availability
                inventory[medicine] -= quantity
                available_vehicles -= 1
                
            elif available_stock >= quantity and available_vehicles == 0:
                # Have stock but no vehicles
                pending_orders.append({
                    **order,
                    'status': 'PENDING',
                    'reason': 'No delivery vehicles available',
                    'estimated_fulfillment': 'Next delivery cycle'
                })
                
            elif available_stock < quantity:
                # Insufficient stock - partial fulfillment
                if available_stock > 0 and available_vehicles > 0:
                    fulfillment = {
                        **order,
                        'status': 'PARTIAL',
                        'allocated_quantity': available_stock,
                        'requested_quantity': quantity,
                        'shortage': quantity - available_stock,
                        'fulfillment_time': datetime.now().isoformat()
                    }
                    fulfilled_orders.append(fulfillment)
                    inventory[medicine] = 0
                    available_vehicles -= 1
                else:
                    pending_orders.append({
                        **order,
                        'status': 'PENDING',
                        'reason': 'Insufficient inventory',
                        'available_stock': available_stock
                    })
        
        # Calculate metrics
        fulfillment_rate = (len(fulfilled_orders) / len(orders) * 100) if orders else 0
        
        return {
            'prioritized_orders': prioritized_orders,
            'fulfilled_orders': fulfilled_orders,
            'pending_orders': pending_orders,
            'metrics': {
                'total_orders': len(orders),
                'fulfilled_count': len(fulfilled_orders),
                'pending_count': len(pending_orders),
                'fulfillment_rate': round(fulfillment_rate, 1),
                'vehicles_used': delivery_capacity - available_vehicles,
                'vehicles_available': available_vehicles
            },
            'inventory_status': self._get_inventory_status(inventory),
            'recommendations': self._get_recommendations(fulfilled_orders, pending_orders, inventory)
        }
    
    def _calculate_priority_score(
        self,
        requester_strain: float,  # 0-100 (Hospital HSI or Pharmacy urgency)
        medicine: str,
        urgency: str,
        quantity: int
    ) -> float:
        """
        Calculate Priority Score using weighted formula
        
        PS = (Requester_Strain * 0.4) + (Medicine_Criticality * 0.3) + (Urgency * 0.3)
        """
        
        # Normalize requester strain (0-100)
        strain_score = min(100, max(0, requester_strain))
        
        # Get medicine criticality (0-100)
        criticality_score = self.MEDICINE_CRITICALITY.get(
            medicine.lower(), 
            self.MEDICINE_CRITICALITY['default']
        )
        
        # Get urgency weight (0-100)
        urgency_score = self.URGENCY_WEIGHTS.get(urgency.upper(), 25)
        
        # Calculate weighted priority score
        priority_score = (
            strain_score * 0.4 +
            criticality_score * 0.3 +
            urgency_score * 0.3
        )
        
        return round(priority_score, 2)
    
    def _get_inventory_status(self, inventory: Dict[str, int]) -> Dict:
        """Get current inventory health status"""
        if not inventory:
            return {"status": "EMPTY", "critical_items": []}
        
        total_items = len(inventory)
        low_stock = [item for item, qty in inventory.items() if qty < 100]
        critical_stock = [item for item, qty in inventory.items() if qty < 50]
        
        if len(critical_stock) > total_items * 0.3:
            status = "CRITICAL"
        elif len(low_stock) > total_items * 0.5:
            status = "LOW"
        elif len(low_stock) > total_items * 0.3:
            status = "MODERATE"
        else:
            status = "HEALTHY"
        
        return {
            "status": status,
            "total_items": total_items,
            "low_stock_count": len(low_stock),
            "critical_stock_count": len(critical_stock),
            "critical_items": critical_stock[:5]  # Top 5 critical items
        }
    
    def _get_recommendations(
        self, 
        fulfilled: List[Dict], 
        pending: List[Dict],
        inventory: Dict[str, int]
    ) -> List[str]:
        """Generate actionable recommendations"""
        recommendations = []
        
        if fulfilled:
            urgent_fulfilled = [o for o in fulfilled if o.get('urgency') == 'URGENT']
            if urgent_fulfilled:
                recommendations.append(f"✅ {len(urgent_fulfilled)} URGENT orders fulfilled immediately")
            recommendations.append(f"📦 Total {len(fulfilled)} orders dispatched")
        
        if pending:
            recommendations.append(f"⏳ {len(pending)} orders pending - requires attention")
            
            no_vehicle = [o for o in pending if 'vehicle' in o.get('reason', '').lower()]
            if no_vehicle:
                recommendations.append(f"🚚 {len(no_vehicle)} orders waiting for delivery vehicles")
            
            no_stock = [o for o in pending if 'inventory' in o.get('reason', '').lower()]
            if no_stock:
                recommendations.append(f"📦 {len(no_stock)} orders pending due to low stock - restock needed")
        
        # Inventory warnings
        critical_items = [item for item, qty in inventory.items() if qty < 50]
        if critical_items:
            recommendations.append(f"⚠️ RESTOCK ALERT: {len(critical_items)} items critically low: {', '.join(critical_items[:3])}")
        
        if not recommendations:
            recommendations.append("✅ All operations normal. Inventory levels healthy.")
        
        return recommendations
